### PR TITLE
KAFKA-17633: Add share group record formatter and parser.

### PR DESCRIPTION
**Pull request by smjn**
As part of KIP-932, a new internal topic `__share_group_state` was introduced. There are 2 types of records which are currently being added in this topic - `ShareSnapshotKey/Value` and `ShareUpdateKey/Value`
In light of this, we must make the existing tooling like `kafka-console-consumer` and `kafka-dump-log` aware of these records for debugging and introspection purposes.
This PR introduces `ShareGroupStateMessageFormatter` to be used used with `kafka-console-consumer` and adds an internal class `ShareGroupStateMessageParser` in `DumpLogSegments.scala`.
Unit tests have been added to `DumpLogSegmentsTest.scala`

### Testing
**kafka-console-consumer**
```
➜ bin/kafka-console-consumer.sh --bootstrap-server localhost:9092 --from-beginning  --topic __share_group_state --property print.key=true   --property print.timestamp=true --property print.headers=true --property print.partition=true --timeout-ms 6000 --formatter org.apache.kafka.tools.consumer.group.share.ShareGroupStateMessageFormatter
{"key":{"version":0,"data":{"groupId":"gs1","topicId":"Uj5wn_FqTXirEASvVZRY1w","partition":0}},"value":{"version":0,"data":{"snapshotEpoch":0,"stateEpoch":0,"leaderEpoch":0,"startOffset":0,"stateBatches":[{"firstOffset":0,"lastOffset":4,"deliveryState":2,"deliveryCount":1}]}}}{"key":{"version":1,"data":{"groupId":"gs1","topicId":"Uj5wn_FqTXirEASvVZRY1w","partition":0}},"value":{"version":0,"data":{"snapshotEpoch":0,"leaderEpoch":0,"startOffset":5,"stateBatches":[{"firstOffset":5,"lastOffset":7,"deliveryState":2,"deliveryCount":1}]}}}
```
**kafka-dump-log**
```
➜ bin/kafka-dump-log.sh --files /tmp/kraft-combined-logs/__share_group_state-34/00000000000000000000.log --share-group-state-decoder
Dumping /tmp/kraft-combined-logs/__share_group_state-34/00000000000000000000.log
Log starting offset: 0
baseOffset: 0 lastOffset: 0 count: 1 baseSequence: 0 lastSequence: 0 producerId: -1 producerEpoch: -1 partitionLeaderEpoch: 0 isTransactional: false isControl: false deleteHorizonMs: OptionalLong.empty position: 0 CreateTime: 1728629394220 size: 138 magic: 2 compresscodec: none crc: 2528359694 isvalid: true
| offset: 0 CreateTime: 1728629394220 keySize: 27 valueSize: 42 sequence: 0 headerKeys: [] key: {"type":"0","data":{"groupId":"gs1","topicId":"Uj5wn_FqTXirEASvVZRY1w","partition":0}} payload: {"type":"0","data":{"snapshotEpoch":0,"stateEpoch":0,"leaderEpoch":0,"startOffset":0,"stateBatches":[{"firstOffset":0,"lastOffset":4,"deliveryState":2,"deliveryCount":1}]}}
baseOffset: 1 lastOffset: 1 count: 1 baseSequence: 0 lastSequence: 0 producerId: -1 producerEpoch: -1 partitionLeaderEpoch: 0 isTransactional: false isControl: false deleteHorizonMs: OptionalLong.empty position: 138 CreateTime: 1728629410581 size: 134 magic: 2 compresscodec: none crc: 1581677863 isvalid: true
| offset: 1 CreateTime: 1728629410581 keySize: 27 valueSize: 38 sequence: 0 headerKeys: [] key: {"type":"1","data":{"groupId":"gs1","topicId":"Uj5wn_FqTXirEASvVZRY1w","partition":0}} payload: {"type":"0","data":{"snapshotEpoch":0,"leaderEpoch":0,"startOffset":5,"stateBatches":[{"firstOffset":5,"lastOffset":7,"deliveryState":2,"deliveryCount":1}]}}
```

### CODE REVIEW

File: core/src/main/scala/kafka/tools/DumpLogSegments.scala [hunk 1]
```diff
@@ -628,6 +630,68 @@ object DumpLogSegments {
     }
   }
 
+  // for test visibility
```
Comment (AndrewJSchofield) [inline core/src/main/scala/kafka/tools/DumpLogSegments.scala:633]: I'd put this class alongside the other subclasses of `MessageParser`, such as above line 631.

File: core/src/main/scala/kafka/tools/DumpLogSegments.scala [hunk 2]
```diff
@@ -628,6 +630,68 @@ object DumpLogSegments {
...
+  class ShareGroupStateMessageParser extends MessageParser[String, String] {
+    private val serde = new ShareCoordinatorRecordSerde()
+
+    private def prepareKey(message: Message, version: Short): String = {
+      val messageAsJson = message match {
+        case m: ShareSnapshotKey =>
+          ShareSnapshotKeyJsonConverter.write(m, version)
+        case m: ShareUpdateKey =>
+          ShareUpdateKeyJsonConverter.write(m, version)
+        case _ => throw new IllegalStateException(s"Message key ${message.getClass.getSimpleName} is not supported.")
```
Comment (AndrewJSchofield) [inline core/src/main/scala/kafka/tools/DumpLogSegments.scala]: I'd use UnknownRecordTypeException in this case.

File: core/src/main/scala/kafka/tools/DumpLogSegments.scala [hunk 3]
```diff
@@ -628,6 +630,68 @@ object DumpLogSegments {
...
+    private def jsonString(jsonNode: JsonNode, version: Short): String = {
+      val json = new ObjectNode(JsonNodeFactory.instance)
+      json.set("type", new TextNode(version.toString))
+      json.set("data", jsonNode)
+      json.toString
+    }
+
+    override def parse(record: Record): (Option[String], Option[String]) = {
+      if (!record.hasKey)
+        throw new RuntimeException(s"Failed to decode message at offset ${record.offset} using offset " +
```
Comment (AndrewJSchofield) [inline core/src/main/scala/kafka/tools/DumpLogSegments.scala]: This is not the offset topic decoder :)

**Review by AndrewJSchofield — CHANGES_REQUESTED** [review]
Thanks for the PR. Just a handful of comments. When the PR is approved, let me know and I'll update the KIP.

File: core/src/main/scala/kafka/tools/DumpLogSegments.scala (hunk 1 above)
Comment (smjn) [reply core/src/main/scala/kafka/tools/DumpLogSegments.scala:633]: It's already with other MessageParser specializations:
order
```
class OffsetsMessageParser extends MessageParser[String, String]
private class ClusterMetadataLogMessageParser extends MessageParser[String, String]
private class RemoteMetadataLogMessageParser extends MessageParser[String, String] 
class ShareGroupStateMessageParser extends MessageParser[String, String]
```

File: core/src/main/scala/kafka/tools/DumpLogSegments.scala (hunk 2 above)
Comment (smjn) [reply core/src/main/scala/kafka/tools/DumpLogSegments.scala]: `UnknownRecordTypeException` will produce less human readable messages as it only allows recordType (short)  as the constructor argument.
Using `IllegalStateException` we can actually specify the name of the record class and more readable message.

Will update for key

File: core/src/main/scala/kafka/tools/DumpLogSegments.scala (hunk 1 above)
Comment (AndrewJSchofield) [reply core/src/main/scala/kafka/tools/DumpLogSegments.scala:633]: Hmm. I must have got confused in the review tool :)

Comment (AndrewJSchofield) [reply core/src/main/scala/kafka/tools/DumpLogSegments.scala:633]: Yup. I did. Sigh. I need more sleep.

**Review by AndrewJSchofield — APPROVED** [review]

**Review by omkreddy — APPROVED** [review]
@smjn Thanks for the PR. LGTM

**Pull request merged**